## Exploring Ebay Car Sales Data

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

#### Printing existing columns 

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

#### Converting column names to snake case

Changing the column names to Python's preferred snakecase.
3 keyword names were changed to 2 keyword + descriptive names

In [7]:
#column names that change in addition to the snake case change
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

#### Initial exploration and cleaning

Finding:
* Text columns where values are the same (not useful for analysis purposes).
* Numeric data stored as text (clean and convert)

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Columns with mostly one value (candidates to be dropped):
* seller (49999)
* offer_type (49999)

Numeric data as text (needs to be cleaned):
* date_crawled
* price
* registration_year
* power_ps
* registration_month
* ad_created
* postal_code
* nr_of_pictures
* last_seen

Need more investigation:
* odometer (kms / miles / ?)
* brand (look for typos / case changes)


In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Investigating odometer data

In [10]:
autos["odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

Investigating brand data

In [11]:
autos["brand"].value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
smart               701
citroen             701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

In [12]:
#removing $ symbol and ','.  Then convert to float
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
#convert to float
autos["price"] = autos["price"].astype(float)

In [13]:
# odometer remove km and ','. Then convert to int and rename column
autos["odometer"] = autos["odometer"].str.replace("kms","")
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")

#conver to int
autos["odometer"] = autos["odometer"].astype(int)

#rename column to odometer_km
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [14]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

#### Analyzing odometer and price

In [15]:
odometer = autos["odometer_km"]

print (min(odometer))
print (max(odometer))
print (odometer.unique().shape)
print (odometer.describe())
print (odometer.value_counts().sort_index(ascending=False))

5000
150000
(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64


In [16]:
price = autos["price"]

print (min(price))
print (max(price))
print (price.unique().shape)
print (price.describe())
print (price.value_counts().sort_index(ascending=False))
print (price.value_counts().sort_index(ascending=False).head(10))

0.0
99999999.0
(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
              ... 
5.0              2
3.0              1
2.0              3
1.0            156
0.0           1421
Name: price, Length: 2357, dtype: int64
99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
Name: price, dtype: int64


In [17]:
autos_bool_price = autos["price"].between(100,100000)
filtered_autos = autos[autos_bool_price]
print(filtered_autos["price"].describe())

count    48185.000000
mean      5796.099741
std       7525.532405
min        100.000000
25%       1250.000000
50%       3000.000000
75%       7499.000000
max      99900.000000
Name: price, dtype: float64


#### Removing outliers
I'm removing cars with prices above 100k to avoid outliers.  The removed cars represents around 3.6% of the total data.

In [18]:
autos = filtered_autos
print(autos["price"].describe())

count    48185.000000
mean      5796.099741
std       7525.532405
min        100.000000
25%       1250.000000
50%       3000.000000
75%       7499.000000
max      99900.000000
Name: price, dtype: float64


### Exploring date columns

In [19]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [20]:
autos[['date_crawled','ad_created','last_seen']][0:5].dtypes

date_crawled    object
ad_created      object
last_seen       object
dtype: object

In [21]:
date_crawled = autos['date_crawled']
date_crawled = date_crawled.str[:10]
data_crawled = date_crawled.value_counts(normalize=True, dropna=False)
print (data_crawled.sort_index())

2016-03-05    0.025340
2016-03-06    0.014050
2016-03-07    0.036090
2016-03-08    0.033164
2016-03-09    0.033019
2016-03-10    0.032313
2016-03-11    0.032624
2016-03-12    0.036920
2016-03-13    0.015690
2016-03-14    0.036692
2016-03-15    0.034305
2016-03-16    0.029470
2016-03-17    0.031504
2016-03-18    0.012867
2016-03-19    0.034762
2016-03-20    0.037813
2016-03-21    0.037190
2016-03-22    0.032811
2016-03-23    0.032292
2016-03-24    0.029449
2016-03-25    0.031504
2016-03-26    0.032292
2016-03-27    0.031109
2016-03-28    0.034949
2016-03-29    0.034139
2016-03-30    0.033703
2016-03-31    0.031856
2016-04-01    0.033662
2016-04-02    0.035633
2016-04-03    0.038601
2016-04-04    0.036567
2016-04-05    0.013054
2016-04-06    0.003175
2016-04-07    0.001390
Name: date_crawled, dtype: float64


In [22]:
ad_created = autos['ad_created']
ad_created = ad_created.str[:10]
ad_created = ad_created.value_counts(normalize=True, dropna=False)
print (ad_created.sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038850
2016-04-04    0.036920
2016-04-05    0.011788
2016-04-06    0.003258
2016-04-07    0.001245
Name: ad_created, Length: 76, dtype: float64


In [23]:
last_seen = autos['last_seen']
last_seen = last_seen.str[:10]
last_seen = last_seen.value_counts(normalize=True, dropna=False)
print (last_seen.sort_index())

2016-03-05    0.001079
2016-03-06    0.004317
2016-03-07    0.005437
2016-03-08    0.007326
2016-03-09    0.009567
2016-03-10    0.010646
2016-03-11    0.012411
2016-03-12    0.023804
2016-03-13    0.008882
2016-03-14    0.012639
2016-03-15    0.015876
2016-03-16    0.016437
2016-03-17    0.028121
2016-03-18    0.007305
2016-03-19    0.015773
2016-03-20    0.020650
2016-03-21    0.020546
2016-03-22    0.021376
2016-03-23    0.018574
2016-03-24    0.019736
2016-03-25    0.019114
2016-03-26    0.016623
2016-03-27    0.015544
2016-03-28    0.020836
2016-03-29    0.022310
2016-03-30    0.024717
2016-03-31    0.023846
2016-04-01    0.022870
2016-04-02    0.024883
2016-04-03    0.025132
2016-04-04    0.024551
2016-04-05    0.124935
2016-04-06    0.221999
2016-04-07    0.132137
Name: last_seen, dtype: float64


In [24]:
autos["registration_year"].describe()

count    48185.000000
mean      2004.730456
std         87.932039
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

According to the previous data, there is a min registration year of 1000 and a max of 9999 which are out of the valid range. 
Year 1k was long time ago and I'd need a time machine to check if cars exist in 9999 :-)

Assuming a minimum year of 1900 and a max of 2016 (dataset year), the valid car range would be 1900 to 2016.

Below, I'm checking how many cars are outside of the valid registration year range.

In [25]:
bool_reg_year = autos["registration_year"].between(1900,2016)

autos_in_range = autos.loc[bool_reg_year,:]

top_15=autos_in_range["registration_year"].value_counts(normalize=True).sort_values(ascending=False)[:15]
print(top_15)
#normalize = True
#With normalize set to True, 
#returns the relative frequency by dividing all values 
#by the sum of values.
#ref: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
print(sum(top_15))

2000    0.067001
2005    0.062855
1999    0.062164
2004    0.058277
2003    0.058148
2006    0.057608
2001    0.056744
2002    0.053484
1998    0.050461
2007    0.049058
2008    0.047676
2009    0.044912
1997    0.041543
2011    0.034915
2010    0.034180
Name: registration_year, dtype: float64
0.7790253276617797


From the data above, 77.9% of the vehicles were registered from 1997 to 2011

#### Exploring price by brand


In [26]:
autos["brand"].value_counts(normalize=True).sort_values(ascending=False)

volkswagen        0.213158
bmw               0.108748
opel              0.108395
mercedes_benz     0.095984
audi              0.086043
ford              0.069627
renault           0.047857
peugeot           0.029470
fiat              0.025983
seat              0.018948
skoda             0.016084
nissan            0.015337
mazda             0.015274
smart             0.014320
citroen           0.014154
toyota            0.012680
hyundai           0.009941
sonstige_autos    0.009339
volvo             0.009028
mini              0.008654
mitsubishi        0.008177
honda             0.008032
kia               0.007118
alfa_romeo        0.006641
suzuki            0.005894
chevrolet         0.005645
porsche           0.005271
chrysler          0.003487
dacia             0.002677
daihatsu          0.002511
jeep              0.002241
subaru            0.002096
land_rover        0.002055
saab              0.001640
daewoo            0.001557
jaguar            0.001494
trabant           0.001370
r

In [27]:
top_10_brands = autos["brand"].value_counts(normalize=True).sort_values(ascending=False)[:10]
print(top_10_brands)
print(sum(top_10_brands))

volkswagen       0.213158
bmw              0.108748
opel             0.108395
mercedes_benz    0.095984
audi             0.086043
ford             0.069627
renault          0.047857
peugeot          0.029470
fiat             0.025983
seat             0.018948
Name: brand, dtype: float64
0.8042129293348553


80% of the vehicles are in the top 10 brands of the list.
I'm focusing on those 10 first to get the average price by brand

In [28]:
price_mean_by_brand = {}

for brand in top_10_brands.index:
    current_brand= autos["brand"]== brand
    mean = autos.loc[current_brand,"price"].mean()
    price_mean_by_brand[brand] = int(mean)
    
price_mean_by_brand


{'volkswagen': 5366,
 'bmw': 8181,
 'opel': 2974,
 'mercedes_benz': 8484,
 'audi': 9219,
 'ford': 3719,
 'renault': 2450,
 'peugeot': 3086,
 'fiat': 2815,
 'seat': 4348}

In [29]:
dict(sorted(price_mean_by_brand.items(), key=lambda item: item[1]))

{'renault': 2450,
 'fiat': 2815,
 'opel': 2974,
 'peugeot': 3086,
 'ford': 3719,
 'seat': 4348,
 'volkswagen': 5366,
 'bmw': 8181,
 'mercedes_benz': 8484,
 'audi': 9219}

* Audi, MB and Audi are above 8k, those are the more expensive cars in avg.
* While VW at 5.3k avg price (VW has the largest share ~21.31%)
* renault, fiat and opel are among the less expensive cars ( < 3k )
* Ford is one of the non european top 6 cars, near to the less expensive cars( 3.7k ) and ranked in 6th place on the brand popularity rank.

#### Exploring car mileage


In [31]:
#Calculating mean mileage
mileage_mean_by_brand = {}

for brand in top_10_brands.index:
    current_brand = autos["brand"]== brand
    mean = autos.loc[current_brand,"odometer_km"].mean()
    mileage_mean_by_brand[brand] = int(mean)
    
mileage_mean_by_brand

{'volkswagen': 129008,
 'bmw': 132884,
 'opel': 129442,
 'mercedes_benz': 131088,
 'audi': 129634,
 'ford': 124383,
 'renault': 128261,
 'peugeot': 127316,
 'fiat': 117408,
 'seat': 122020}

In [42]:
dict(sorted(mileage_mean_by_brand.items(), key=lambda item: item[1]))


{'fiat': 117408,
 'seat': 122020,
 'ford': 124383,
 'peugeot': 127316,
 'renault': 128261,
 'volkswagen': 129008,
 'opel': 129442,
 'audi': 129634,
 'mercedes_benz': 131088,
 'bmw': 132884}

In [47]:
bpm_mean = pd.Series(price_mean_by_brand)
bpm_mean = bpm_mean.sort_values(ascending=False)
bpm_df = pd.DataFrame(bpm_mean, columns=['mean_price'])
bpm_df

mean_price
audi                 9219
mercedes_benz        8484
bmw                  8181
volkswagen           5366
seat                 4348
ford                 3719
peugeot              3086
opel                 2974
fiat                 2815
renault              2450

In [48]:
mmb_mean = pd.Series(mileage_mean_by_brand)
mmb_mean = mmb_mean.sort_values(ascending=False)
mmb_df = pd.DataFrame(mmb_mean, columns=['mean_odometer_km'])
mmb_df

mean_odometer_km
bmw                      132884
mercedes_benz            131088
audi                     129634
opel                     129442
volkswagen               129008
renault                  128261
peugeot                  127316
ford                     124383
seat                     122020
fiat                     117408

In [51]:
results_df = bpm_df
results_df["mean_odometer"]=mmb_mean
results_df

mean_price  mean_odometer
audi                 9219         129634
mercedes_benz        8484         131088
bmw                  8181         132884
volkswagen           5366         129008
seat                 4348         122020
ford                 3719         124383
peugeot              3086         127316
opel                 2974         129442
fiat                 2815         117408
renault              2450         128261

When comparing prices and avg kms(odometer), cars with higher prices are almost also the cars with more kms.  This don't necesary means that those cars lasts more, it could also be an indicator that luxury car owners tend to use their cars for more time. 